pkg/bytecode/code.go

```diff
@@ -0,0 +1,96 @@
+package bytecode
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+)
+
+const (
+	// OpConstant defines a constant that will be referred to by index
+	// in the bytecode.
+	OpConstant Opcode = iota
+	// OpGetGlobal retrieves a symbol from the symbol table at the
+	// specified index.
+	OpGetGlobal
+	// OpSetGlobal adds a symbol to the specified index in the symbol
+	// table.
+	OpSetGlobal
+)
+
+// ErrUnknownOp is returned when an unknown opcode is encountered.
+var ErrUnknownOp = errors.New("unknown opcode")
+
+// definitions is a mapping of OpCode to OpDefinition.
+var definitions = map[Opcode]*OpDefinition{
+	// The definition for OpConstant says that its only operand is two
+	// bytes wide, which makes it a uint16. This means that an evy
+	// bytecode program can only have 65535 constants defined.
+	OpConstant:  {"OpConstant", []int{2}},
+	OpGetGlobal: {"OpGetGlobal", []int{2}},
+	OpSetGlobal: {"OpSetGlobal", []int{2}},
+}
+
+// OpDefinition defines a name and expected operand width for each OpCode.
+type OpDefinition struct {
+	Name          string
+	OperandWidths []int
+}
+
+// Opcode defines the type of operation to be performed when reading
+// the bytecode.
+type Opcode byte
+
+// Lookup returns an OpDefinition for the value of an Opcode, any
+// unknown value will result in an error.
+func Lookup(op byte) (*OpDefinition, error) {
+	def, ok := definitions[Opcode(op)]
+	if !ok {
+		return nil, fmt.Errorf("%w: %d", ErrUnknownOp, op)
+	}
+	return def, nil
+}
+
+// Make assembles a single bytecode instruction out of an Opcode and
+// an optional list of operands.
+func Make(op Opcode, operands ...int) ([]byte, error) {
+	def, ok := definitions[op]
+	if !ok {
```

Comment:
arguably you could/should use Lookup(op) for this? (keep errors in sync)

Reply:
Yes good point, The book avoids this as a trade off for an "error free" Make function.

> Note that we don’t use the Lookup function to get to the definition, which gives us a much more usable function signature for Make in the tests later on. By circumventing Lookup and not having to return possible errors, we can use Make to easily build up bytecode instructions without having to check for errors after every call. The risk of producing empty byte slices by using an unknown opcode is one we’re willing to take,

But since we return an error now anyway it makes sense to use Lookup.